actionpack/lib/action_dispatch/http/request.rb

```diff
@@ -199,6 +199,23 @@ def headers
       @headers ||= Http::Headers.new(self)
     end
 
+    # Early Hints is an H2 status code that indicates hints to help a client start
+    # making preparations for processing the final response.
+    #
+    # If the env contains +rack.early_hints+ then the server accepts HTTP2 push for Link headers.
+    #
+    # The +send_early_hints+ method accepts an array of links as follows:
+    #
+    #   send_early_hints({ link: "path/to/stylesheet.css", as: "stylesheet}, { link: "path/to/javascript.js", as: "script"})
+    #
+    # If you are using +javascript_include_tag+ or +stylesheet_link_tag+ the
+    # Early Hints headers are included by default if supported.
+    def send_early_hints(links)
+      return unless env["rack.early_hints"]
+
+      env["rack.early_hints"].call(links)
```

Comment:
could avoid calling doing the same lookup twice:

```Ruby
env["rack.early_hints"]&.call(links)
```

or

```Ruby
callback = env["rack.early_hints"]
callback.call(links) if callback

Reply:
It's a hash lookup. I'm sure it'll be fine.

Reply:
I've considered this but agree with @matthewd.

Drying things up isn't a benefit if the code is harder to read and understand and the performance impact of calling a hash lookup twice will be negligible.

Reply:
I was just curious about the performance impact, so I did a quick benchmarking.

### TL;DR

> ... the performance impact of calling a hash lookup twice will be negligible.

### Details

```
# 1.18x slower than v2 if env["rack.early_hints"]
def v1
  return unless env["rack.early_hints"]
  env["rack.early_hints"].call(links)
end

# 1.09x slower than v1 IF NOT env["rack.early_hints"]
def v2
  callback = env["rack.early_hints"]
  callback.call(links) if callback
end
```
<details>
  <summary>Click to show code</summary>

    #!/usr/bin/env ruby

    require 'benchmark/ips'
    require 'date'

    LOOKUP = { "foo" => ->() {} }

    def happy_case
      return unless LOOKUP["foo"]

      LOOKUP["foo"].call()
    end

    def happy_case_with_assignment
      callback = LOOKUP["foo"]

      callback.call() if callback
    end

    def alternase_case
      return unless LOOKUP["bar"]

      LOOKUP["bar"].call()
    end

    def alternase_case_with_assignment
      callback = LOOKUP["bar"]

      callback.call() if callback
    end


    # == Result
    #
    # happy_case: 1.18x slower
    Benchmark.ips do |x|
      # Configure the number of seconds used during
      # the warmup phase (default 2) and calculation phase (default 5)
      x.config(time: 5, warmup: 2)

      x.report("happy_case") { happy_case }
      x.report("happy_case_with_assignment") { happy_case_with_assignment }

      # Compare the iterations per second of the various reports!
      x.compare!
    end

    # == Result
    # 
    # alternase_case_with_assignment: 1.09x slower
    Benchmark.ips do |x|
      # Configure the number of seconds used during
      # the warmup phase (default 2) and calculation phase (default 5)
      x.config(time: 5, warmup: 2)

      x.report("alternase_case") { alternase_case }
      x.report("alternase_case_with_assignment") { alternase_case_with_assignment }

      # Compare the iterations per second of the various reports!
      x.compare!
    end

    __END__

    Calculating -------------------------------------
              happy_case   111.305k i/100ms
    happy_case_with_assignment
                           127.680k i/100ms
    -------------------------------------------------
              happy_case      3.704M (± 2.1%) i/s -     18.588M
    happy_case_with_assignment
                              4.353M (± 1.2%) i/s -     21.833M

    Comparison:
    happy_case_with_assignment:  4353466.9 i/s
              happy_case:  3704451.9 i/s - 1.18x slower

    Calculating -------------------------------------
          alternase_case   146.388k i/100ms
    alternase_case_with_assignment
                           154.871k i/100ms
    -------------------------------------------------
          alternase_case      7.631M (± 2.9%) i/s -     38.207M
    alternase_case_with_assignment
                              6.977M (± 6.8%) i/s -     34.846M

    Comparison:
          alternase_case:  7630584.3 i/s
    alternase_case_with_assignment:  6977239.4 i/s - 1.09x slower

</details>

Reply:
interesting, would have been nice to see how the `&.` syntax compares.